File: bindings/java/src/test/java/io/angzarr/router/conformance/saga/SagaSteps.java
package io.angzarr.router.conformance.saga;

import io.angzarr.router.conformance.Builders;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import io.angzarr.SagaHandleRequest;
import io.angzarr.SagaResponse;
import io.angzarr.router.CodedError;
import io.angzarr.router.Router;
import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import java.util.Map;
import test.counter.OrderSagaAngzarr;

/** Step definitions for saga.feature — the OrderSaga translation-side dispatch. */
public class SagaSteps {

  private Router router;
  private SagaResponse resp;
  private CodedError err;

  @Before
  public void before() {
    router = new Router();
    OrderSagaAngzarr.registerOrderSaga(router, new SagaFixture());
    resp = null;
    err = null;
  }

  @After
  public void after() {
    if (router != null) {
      router.close();
    }
  }

  private void dispatch(SagaHandleRequest req) {
    try {
      resp = router.dispatchSaga(req);
      err = null;
    } catch (CodedError e) {
      err = e;
      resp = null;
    }
  }

  @Given("an order saga delivering to {string}")
  public void anOrderSaga(String target) {
    // The fixture is registered in @Before; the delivery target ("inventory")
    // is part of the declaration the generated wiring already carries.
  }

  @When("an Increased event is dispatched with destination inventory sequence {int}")
  public void increasedWithDestination(int seq) {
    dispatch(Builders.sagaEventSource("test.counter.Increased", Map.of("inventory", seq)));
  }

  @When("a Reserve event is dispatched")
  public void reserveEvent() {
    dispatch(Builders.sagaEventSource("test.counter.Reserve", null));
  }

  @When("a source with no pages is dispatched")
  public void sourceNoPages() {
    dispatch(Builders.sagaSourceNoPages());
  }

  @When("a request with no source is dispatched")
  public void requestNoSource() {
    dispatch(Builders.sagaRequestNoSource());
  }

  @When("a rejection of Reserve is dispatched")
  public void rejectionReserve() {
    dispatch(Builders.sagaRejectionSource("test.counter.Reserve"));
  }

  @When("a rejection of Unwatched is dispatched")
  public void rejectionUnwatched() {
    dispatch(Builders.sagaRejectionSource("test.counter.Unwatched"));
  }

  @Then("the saga emits one command to {string}")
  public void emitsOneCommand(String target) {
    assertNull(err, "dispatch failed");
    assertEquals(1, resp.getCommandsList().size(), "emitted commands");
    assertEquals(target, resp.getCommandsList().get(0).getCover().getDomain(), "command target");
  }

  @Then("the command carries destination sequence {int}")
  public void commandCarriesSequence(int seq) {
    assertEquals(seq, resp.getCommandsList().get(0).getPages(0).getHeader().getSequence());
  }

  @Then("the saga emits no commands")
  public void emitsNoCommands() {
    assertNull(err, "dispatch failed");
    assertEquals(0, resp.getCommandsList().size(), "expected no commands");
  }

  @Then("the dispatch fails with {word}")
  public void dispatchFailsWith(String code) {
    assertNotNull(err, "expected coded error " + code);
    assertEquals(code, err.code);
  }

  @Then("the saga injects one fact event")
  public void injectsOneEvent() {
    assertNull(err, "dispatch failed");
    assertEquals(1, resp.getEventsList().size(), "injected events");
  }

  @Then("the saga injects no events")
  public void injectsNoEvents() {
    assertNull(err, "dispatch failed");
    assertEquals(0, resp.getEventsList().size(), "expected no events");
  }
}

File: bindings/csharp/conformance/Saga/SagaSteps.cs
using System.Collections.Generic;
using Angzarr;
using Angzarr.Router;
using NUnit.Framework;
using Reqnroll;
using TC = Test.Counter;

namespace Angzarr.Router.Conformance.Saga;

/// <summary>Step definitions for saga.feature — the OrderSaga translation-side
/// dispatch. Scoped to the feature (saga and pm share step text).</summary>
[Binding]
[Scope(Feature = "Order saga dispatch")]
public sealed class SagaSteps
{
    private Router _router = null!;
    private SagaResponse? _resp;
    private CodedError? _err;

    [BeforeScenario]
    public void Before()
    {
        _router = new Router();
        TC.counter_angzarr.RegisterOrderSaga(_router, new SagaFixture());
        _resp = null;
        _err = null;
    }

    [AfterScenario]
    public void After() => _router?.Dispose();

    private void Dispatch(SagaHandleRequest req)
    {
        try
        {
            _resp = _router.DispatchSaga(req);
            _err = null;
        }
        catch (CodedError e)
        {
            _err = e;
            _resp = null;
        }
    }

    [Given("an order saga delivering to {string}")]
    public void AnOrderSaga(string target)
    {
        // The fixture is registered in Before; the delivery target ("inventory")
        // is part of the declaration the generated wiring already carries.
    }

    [When("an Increased event is dispatched with destination inventory sequence {int}")]
    public void IncreasedWithDestination(int seq) =>
        Dispatch(
            Builders.SagaEventSource(
                "test.counter.Increased",
                new Dictionary<string, uint> { ["inventory"] = (uint)seq }
            )
        );

    [When("a Reserve event is dispatched")]
    public void ReserveEvent() => Dispatch(Builders.SagaEventSource("test.counter.Reserve", null));

    [When("a source with no pages is dispatched")]
    public void SourceNoPages() => Dispatch(Builders.SagaSourceNoPages());

    [When("a request with no source is dispatched")]
    public void RequestNoSource() => Dispatch(Builders.SagaRequestNoSource());

    [When("a rejection of Reserve is dispatched")]
    public void RejectionReserve() =>
        Dispatch(Builders.SagaRejectionSource("test.counter.Reserve"));

    [When("a rejection of Unwatched is dispatched")]
    public void RejectionUnwatched() =>
        Dispatch(Builders.SagaRejectionSource("test.counter.Unwatched"));

    [Then("the saga emits one command to {string}")]
    public void EmitsOneCommand(string target)
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Commands.Count, Is.EqualTo(1), "emitted commands");
        Assert.That(_resp.Commands[0].Cover.Domain, Is.EqualTo(target), "command target");
    }

    [Then("the command carries destination sequence {int}")]
    public void CommandCarriesSequence(int seq) =>
        Assert.That((int)_resp!.Commands[0].Pages[0].Header.Sequence, Is.EqualTo(seq));

    [Then("the saga emits no commands")]
    public void EmitsNoCommands()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Commands.Count, Is.EqualTo(0), "expected no commands");
    }

    [Then("the dispatch fails with {word}")]
    public void DispatchFailsWith(string code)
    {
        Assert.That(_err, Is.Not.Null, "expected coded error " + code);
        Assert.That(_err!.Code, Is.EqualTo(code));
    }

    [Then("the saga injects one fact event")]
    public void InjectsOneEvent()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Events.Count, Is.EqualTo(1), "injected events");
    }

    [Then("the saga injects no events")]
    public void InjectsNoEvents()
    {
        Assert.That(_err, Is.Null, "dispatch unexpectedly failed");
        Assert.That(_resp!.Events.Count, Is.EqualTo(0), "expected no events");
    }
}
